src/components/database/updateReviewerStats.ts

```diff
@@ -0,0 +1,231 @@
+/**
+ * Function to update reviewer/assigner stats:
+ * - the "reviewer_list" and "assigner_list" on application
+ * - the "application_reviewer_action" table, which contains a list of every
+ *   reviewer's possible actions (Assign, Review, Re-assign, etc) for every
+ *   application they could be interacting with
+ *
+ * It is called by either:
+ * - a notification listener (in postgresConnect), which is notified by triggers
+ *   on various tables (see the cases in `updateReviewerStatsFromDBEvent`
+ *   below).
+ * - the "generateReviewAssignments" Action
+ */
+
+import DBConnect from '../../components/database/databaseConnect'
+import { errorMessage } from '../utilityFunctions'
+
+const isManualUpdate: Boolean = process.argv[2] === '--update-reviewer-stats'
+
+interface NotificationPayload {
+  tableName: string
+  data: Record<string, unknown>
+  operation: 'UPDATE' | 'INSERT'
+}
+
+export const updateReviewerStats = async (applicationId: number, userIds?: number[]) => {
+  const db = databaseMethods
+
+  let reviewerIds = userIds ?? (await db.getAllStaffForApplication(applicationId))
+
+  // Update reviewer/assigner lists on application table
+  console.log('Updating reviewer/assigner lists for application:', applicationId)
+  await db.updateApplicationStaffLists(applicationId)
+
+  // Update application_reviewer_actions
+  reviewerIds = reviewerIds.filter((id) => id !== null)
+
+  for (const userId of reviewerIds) {
+    console.log(`Updating actions for user ${userId} on application ${applicationId}`)
+    await db.upsertReviewerAction(applicationId, userId)
+  }
+
+  // Clean up NULL records
+  await db.cleanupNullReviewerActions(applicationId)
+}
+
+// When triggered by a Database event, this preliminary function gets the
+// applicationId and the list of reviewers to update. It is different for each
+// trigger case, so this function can standardise the input required by
+// `updateReviewerStats`, which does the actual updates
+export const updateReviewerStatsFromDBEvent = async ({
+  tableName,
+  data,
+  operation,
+}: NotificationPayload) => {
+  let applicationId: number = 0
+  let reviewerIds: number[] = []
+
+  const db = databaseMethods
+
+  console.log(
+    `\nUpdating reviewer action stats due to ${operation} trigger on table: ${tableName} `
+  )
+
+  switch (tableName) {
+    case 'review_assignment':
+      applicationId = data.application_id as number
+      // Update ALL staff by leaving reviewerIds empty
+      break
+    case 'review_status_history':
+      const { application_id, reviewer_id, assigner_id } = await db.getApplicationDataFromReview(
+        data.review_id as number
+      )
+      applicationId = application_id
+      reviewerIds.push(...new Set([reviewer_id, assigner_id]))
+      break
+    case 'application':
+      applicationId = data.id as number
+      // Update ALL staff by leaving reviewerIds empty
+      break
```

Comment:
default case would make applicationId = 0 ?

Reply:
Yeah, which would do nothing (tested the sql, doesn't throw any errors, but updates no records)